# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: editor/libeditor/HTMLEditHelpers.h
# Commit: 554d63dad12e
# Full Hash: 554d63dad12e4d1f2560e99de1fafac9a97da026
# Author: Masayuki Nakano <masayuki@d-toybox.com>
# Date: 2023-09-27 09:02:41
# Regressor Bug: 1851951
# File Overlap Count: 1
# Description:
#   Bug 1851951 - Make `HTMLEditor` refer computed `display` instead of the HTML default style at considering block or inline element r=m_kato
#   
#   Blink [1] and WebKit [2] refers `display-outside` value when they consider whether an
#   element is a block or an inline.
#   
# ==============================================================================

diff -r 97dc7be3b93c -r 554d63dad12e editor/libeditor/HTMLEditHelpers.h
--- a/editor/libeditor/HTMLEditHelpers.h	Wed Sep 27 02:50:33 2023 +0000
+++ b/editor/libeditor/HTMLEditHelpers.h	Wed Sep 27 03:01:26 2023 +0000
@@ -15,7 +15,6 @@
 #include "EditorDOMPoint.h"
 #include "EditorForwards.h"
 #include "EditorUtils.h"  // for CaretPoint
-#include "HTMLEditHelpers.h"
 #include "JoinSplitNodeDirection.h"
 
 #include "mozilla/AlreadyAddRefed.h"
@@ -40,6 +39,41 @@
 
 namespace mozilla {
 
+enum class BlockInlineCheck : uint8_t {
+  // BlockInlineCheck is not expected by the root caller.
+  Unused,
+  // Refer only the HTML default style at considering whether block or inline.
+  // All non-HTML elements are treated as inline.
+  UseHTMLDefaultStyle,
+  // Refer the element's computed style of display-outside at considering
+  // whether block or inline.
+  // FYI: If editor.block_inline_check.use_computed_style pref is set to false,
+  // this is same as HTMLDefaultStyle.
+  UseComputedDisplayOutsideStyle,
+  // Refer the element's computed style of display at considering whether block
+  // or inline.  I.e., this is a good value to look for any block boundary.
+  // E.g., this is proper value when:
+  // * Checking visibility of collapsible white-spaces or <br>
+  // * Looking for whether a padding <br> is required
+  // * Looking for a caret position
+  // FYI: If editor.block_inline_check.use_computed_style pref is set to false,
+  // this is same as HTMLDefaultStyle.
+  UseComputedDisplayStyle,
+};
+
+/**
+ * Even if the caller wants block boundary caused by display-inline: flow-root
+ * like inline-block, because it's required only when scanning from in it.
+ * I.e., if scanning needs to go to siblings, we don't want to treat
+ * inline-block siblings as inline.
+ */
+[[nodiscard]] inline BlockInlineCheck IgnoreInsideBlockBoundary(
+    BlockInlineCheck aBlockInlineCheck) {
+  return aBlockInlineCheck == BlockInlineCheck::UseComputedDisplayStyle
+             ? BlockInlineCheck::UseComputedDisplayOutsideStyle
+             : aBlockInlineCheck;
+}
+
 enum class WithTransaction { No, Yes };
 inline std::ostream& operator<<(std::ostream& aStream,
                                 WithTransaction aWithTransaction) {